src/pygeoprocessing/routing/routing.pyx

```diff
@@ -4273,15 +4282,35 @@ def detect_outlets(d8_flow_dir_raster_path_band, target_outlet_vector_path):
                 if flow_dir == flow_dir_nodata:
                     continue
 
-                # inspect the outflow pixel neighbor
-                flow_dir_n = flow_dir_block[
-                    yi+D8_YOFFSET[flow_dir],
-                    xi+D8_XOFFSET[flow_dir]]
+                is_outlet = 1
+                if d8_flow_dir_mode:
+                    # inspect the outflow pixel neighbor
+                    flow_dir_n = flow_dir_block[
+                        yi+D8_YOFFSET[flow_dir],
+                        xi+D8_XOFFSET[flow_dir]]
+
+                    # if the outflow pixel is outside the raster boundaries or
+                    # is a nodata pixel it must mean xi,yi is an outlet
+                    if flow_dir_n != flow_dir_nodata:
+                        is_outlet = 0
+                else:
+                    # inspect all the outflow pixel neighbors in MFD mode
+                    for n_dir in range(8):
+                        # shift the 0xF mask to the outflow direction and
+                        # test if there's any outflow or not. 0 means nothing
+                        # flows out of that pixel in the ``n_dir`` direction.
+                        if flow_dir&(0xF<<(n_dir*4)) == 0:
+                            continue
+                        flow_dir_n = flow_dir_block[
+                            yi+D8_YOFFSET[n_dir],
+                            xi+D8_XOFFSET[n_dir]]
+                        if flow_dir_n != flow_dir_nodata:
+                            is_outlet = 0
+                            break
 
                 # if the outflow pixel is outside the raster boundaries or

```

Comment:
I thought it was still relevant since we're about to enter the conditional of we're an outlet, but maybe we want to know the conditions of why it was an outlet? Is it okay if I leave it in? Or can you say more about why I should take it out?

Reply:
Ah I think I said that because there's the same comment above on lines 4296-4297. But it does kinda make sense both places.